ivre/db/mongo.py

```diff
@@ -1142,6 +1142,32 @@ def searchscreenshot(port=None, protocol='tcp', service=None, neg=False):
                            'screenshot': {'$exists': not neg}}
         }}
 
+    @staticmethod
+    def searchcpe(value=None, type=None, vendor=None, product=None,
+                  components=None):
+        """Look for a CPE by value (original cpe string), type (a, o or h),
+        vendor, product or components (the part after the column following
+        the product). No argument will just check for cpe existence.
+
+        """
+        if all(arg is None
+               for arg in [value, type, vendor, product, components]):
+            return {"cpes": {"$exists": True}}
+
+        flt = {}
+        fields = [
+            ("value", value),
+            ("type", type),
+            ("vendor", vendor),
+            ("product", product),
+            ("components", components),
+        ]
+
+        for (field_name, field_val) in fields:
+            if field_val is not None:
+                flt["cpes." + field_name] = field_val
+        return flt
```

Comment:
If I understand your code well, you have a mistake here: if I specify at least two values (say `value=yyy` and `cpe_type=xxx`), the method will return `{"cpes.type": "xxx", "cpes.value: "yyy"}`. This is not what you want, because it would match a document with, for example `{"cpes": [{"type": "xxx", "value": "zzz"}, {"type": "zzz", "value": "yyy"}]}`.

Plus I think this code might simplified a bit. Here is a possibility (not tested):

```Python
    flt = dict(
        (field, value)
        for field, value in [
                ("value", value),
                ("type", cpe_type),
                ("vendor", vendor),
                ("product", product),
                ("components", components),
        ]
        if value is not None
    )
    nfields = len(nfields)
    if nfields == 0:
        return {"cpes": {"$exists": True}}
    if nfields == 1:
        field, value = flt.popitem()
        return {"cpes.%s" % field: value}
    return {"cpes": {"$elemMatch": flt}}
```

What do you think?